by_attr helper function won't compile

I noticed that the `by_attr` helper function provided to be used with `serac::Domain` won't compile. When I try to build an executable using it I get this error 
![image](https://github.com/LLNL/serac/assets/99373982/07c96e0e-5745-4822-8373-e9befeabc106)

duplicating error message in text, so this issue can be searched for more easily in the future

```
/usr/workspace/swartz10/lLido-2.0/examples/NotShapeOpt .cpp:282:34: error: call to ‘ofBoundaryElements' is ambiguous
serac::Domain inputRegion = serac::Domain::ofBoundaryElements(smeshPtr, serac::by_attr(3));

/usr/workspace/swartz10/lido-2.0/smith/serac/src/serac/infrastructure/../../serac/numerics/functional/domain.hpp:109:17: note: candidate function
static Domain ofBoundaryElements(const mfem::Mesh& mesh, std::function<bool(std::vector<vec2>, int)> func);

/usr/workspace/swartz10/lido-2.0/smith/serac/src/serac/infrastructure/../../serac/numerics/functional/domain.hpp:112:17: note: candidate function
static Domain ofBoundaryElements(const mfem::Mesh& mesh, std::function<bool(std::vector<vec3>, int)> func);
1 error generated.
```

--- a/src/serac/numerics/functional/domain.hpp
+++ b/src/serac/numerics/functional/domain.hpp
@@ -141,9 +141,10 @@ Domain operator&(const Domain& a, const Domain& b);
 Domain operator-(const Domain& a, const Domain& b);
 
 /// @brief convenience predicate for creating domains by attribute
+template < int dim >
 inline auto by_attr(int value)
 {
-  return [value](auto, int attr) { return attr == value; };
+  return [value](std::vector< tensor<double, dim> >, int attr) { return attr == value; };
 }
 
 }  // namespace serac

--- a/src/serac/numerics/functional/domain.hpp
+++ b/src/serac/numerics/functional/domain.hpp
@@ -141,10 +141,10 @@ Domain operator&(const Domain& a, const Domain& b);
 Domain operator-(const Domain& a, const Domain& b);
 
 /// @brief convenience predicate for creating domains by attribute
-template < int dim >
+template <int dim>
 inline auto by_attr(int value)
 {
-  return [value](std::vector< tensor<double, dim> >, int attr) { return attr == value; };
+  return [value](std::vector<tensor<double, dim> >, int attr) { return attr == value; };
 }
 
 }  // namespace serac

--- a/src/serac/numerics/functional/tests/domain_tests.cpp
+++ b/src/serac/numerics/functional/tests/domain_tests.cpp
@@ -148,8 +148,8 @@ TEST(domain, of_edges)
   }
 
   {
-    constexpr int dim = 2;
-    auto mesh = import_mesh("beam-quad.mesh");
+    constexpr int dim  = 2;
+    auto          mesh = import_mesh("beam-quad.mesh");
     mesh.FinalizeQuadMesh(true);
     Domain d0 = Domain::ofEdges(mesh, std::function([](std::vector<vec2> x, int /* bdr_attr */) {
                                   return (0.5 * (x[0][0] + x[1][0])) < 0.25;  // x coordinate of edge midpoint
@@ -179,9 +179,9 @@ TEST(domain, of_edges)
 TEST(domain, of_faces)
 {
   {
-    constexpr int dim = 3;
-    auto   mesh = import_mesh("onehex.mesh");
-    Domain d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec3> vertices, int /*bdr_attr*/) {
+    constexpr int dim  = 3;
+    auto          mesh = import_mesh("onehex.mesh");
+    Domain        d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec3> vertices, int /*bdr_attr*/) {
                                   return average(vertices)[0] < 0.25;  // x coordinate of face center
                                 }));
     EXPECT_EQ(d0.quad_ids_.size(), 1);
@@ -206,9 +206,9 @@ TEST(domain, of_faces)
   }
 
   {
-    constexpr int dim = 3;
-    auto   mesh = import_mesh("onetet.mesh");
-    Domain d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec3> vertices, int /* bdr_attr */) {
+    constexpr int dim  = 3;
+    auto          mesh = import_mesh("onetet.mesh");
+    Domain        d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec3> vertices, int /* bdr_attr */) {
                                   // accept face if it contains a vertex whose x coordinate is less than 0.1
                                   for (auto v : vertices) {
                                     if (v[0] < 0.1) return true;
@@ -236,9 +236,9 @@ TEST(domain, of_faces)
   }
 
   {
-    constexpr int dim = 2;
-    auto   mesh = import_mesh("beam-quad.mesh");
-    Domain d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec2> vertices, int /* attr */) {
+    constexpr int dim  = 2;
+    auto          mesh = import_mesh("beam-quad.mesh");
+    Domain        d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec2> vertices, int /* attr */) {
                                   return average(vertices)[0] < 2.25;  // x coordinate of face center
                                 }));
     EXPECT_EQ(d0.quad_ids_.size(), 2);
@@ -266,9 +266,9 @@ TEST(domain, of_faces)
 TEST(domain, of_elements)
 {
   {
-    constexpr int dim = 3;
-    auto   mesh = import_mesh("patch3D_tets_and_hexes.mesh");
-    Domain d0   = Domain::ofElements(mesh, std::function([](std::vector<vec3> vertices, int /*bdr_attr*/) {
+    constexpr int dim  = 3;
+    auto          mesh = import_mesh("patch3D_tets_and_hexes.mesh");
+    Domain        d0   = Domain::ofElements(mesh, std::function([](std::vector<vec3> vertices, int /*bdr_attr*/) {
                                      return average(vertices)[0] < 0.7;  // x coordinate of face center
                                    }));
 
@@ -298,9 +298,9 @@ TEST(domain, of_elements)
   }
 
   {
-    constexpr int dim = 2;
-    auto   mesh = import_mesh("patch2D_tris_and_quads.mesh");
-    Domain d0   = Domain::ofElements(
+    constexpr int dim  = 2;
+    auto          mesh = import_mesh("patch2D_tris_and_quads.mesh");
+    Domain        d0   = Domain::ofElements(
         mesh, std::function([](std::vector<vec2> vertices, int /* attr */) { return average(vertices)[0] < 0.45; }));
     EXPECT_EQ(d0.tri_ids_.size(), 1);
     EXPECT_EQ(d0.quad_ids_.size(), 1);

--- a/src/serac/numerics/functional/tests/domain_tests.cpp
+++ b/src/serac/numerics/functional/tests/domain_tests.cpp
@@ -173,6 +173,9 @@ TEST(domain, of_edges)
 
     // check that by_attr compiles
     Domain d4 = Domain::ofEdges(mesh, by_attr<dim>(3));
+
+    Domain d5 = Domain::ofBoundaryElements(mesh, [](std::vector<vec2>, int) { return true; });
+    EXPECT_EQ(d5.edge_ids_.size(), 18);  // 1x8 row of quads has 18 boundary edges
   }
 }
 
@@ -203,6 +206,9 @@ TEST(domain, of_faces)
 
     // check that by_attr compiles
     Domain d4 = Domain::ofFaces(mesh, by_attr<dim>(3));
+
+    Domain d5 = Domain::ofBoundaryElements(mesh, [](std::vector<vec3>, int) { return true; });
+    EXPECT_EQ(d5.quad_ids_.size(), 6);
   }
 
   {
@@ -233,6 +239,9 @@ TEST(domain, of_faces)
 
     // check that by_attr compiles
     Domain d4 = Domain::ofFaces(mesh, by_attr<dim>(3));
+
+    Domain d5 = Domain::ofBoundaryElements(mesh, [](std::vector<vec3>, int) { return true; });
+    EXPECT_EQ(d5.tri_ids_.size(), 4);
   }
 
   {

--- a/src/serac/numerics/functional/tests/domain_tests.cpp
+++ b/src/serac/numerics/functional/tests/domain_tests.cpp
@@ -114,6 +114,10 @@ TEST(domain, of_edges)
     Domain d3 = d0 & d1;
     EXPECT_EQ(d3.edge_ids_.size(), 1);
     EXPECT_EQ(d3.dim_, 1);
+
+    // note: by_attr doesn't apply to edge sets in 3D, since
+    //       mfem doesn't have the notion of edge attributes
+    // Domain d4 = Domain::ofEdges(mesh, by_attr<dim>(3));
   }
 
   {
@@ -137,9 +141,14 @@ TEST(domain, of_edges)
     Domain d3 = d0 & d1;
     EXPECT_EQ(d3.edge_ids_.size(), 1);
     EXPECT_EQ(d3.dim_, 1);
+
+    // note: by_attr doesn't apply to edge sets in 3D, since
+    //       mfem doesn't have the notion of edge attributes
+    // Domain d4 = Domain::ofEdges(mesh, by_attr<dim>(3));
   }
 
   {
+    constexpr int dim = 2;
     auto mesh = import_mesh("beam-quad.mesh");
     mesh.FinalizeQuadMesh(true);
     Domain d0 = Domain::ofEdges(mesh, std::function([](std::vector<vec2> x, int /* bdr_attr */) {
@@ -161,12 +170,16 @@ TEST(domain, of_edges)
     Domain d3 = d0 & d1;
     EXPECT_EQ(d3.edge_ids_.size(), 0);
     EXPECT_EQ(d3.dim_, 1);
+
+    // check that by_attr compiles
+    Domain d4 = Domain::ofEdges(mesh, by_attr<dim>(3));
   }
 }
 
 TEST(domain, of_faces)
 {
   {
+    constexpr int dim = 3;
     auto   mesh = import_mesh("onehex.mesh");
     Domain d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec3> vertices, int /*bdr_attr*/) {
                                   return average(vertices)[0] < 0.25;  // x coordinate of face center
@@ -187,9 +200,13 @@ TEST(domain, of_faces)
     Domain d3 = d0 & d1;
     EXPECT_EQ(d3.quad_ids_.size(), 0);
     EXPECT_EQ(d3.dim_, 2);
+
+    // check that by_attr compiles
+    Domain d4 = Domain::ofFaces(mesh, by_attr<dim>(3));
   }
 
   {
+    constexpr int dim = 3;
     auto   mesh = import_mesh("onetet.mesh");
     Domain d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec3> vertices, int /* bdr_attr */) {
                                   // accept face if it contains a vertex whose x coordinate is less than 0.1
@@ -213,9 +230,13 @@ TEST(domain, of_faces)
     Domain d3 = d0 & d1;
     EXPECT_EQ(d3.tri_ids_.size(), 1);
     EXPECT_EQ(d3.dim_, 2);
+
+    // check that by_attr compiles
+    Domain d4 = Domain::ofFaces(mesh, by_attr<dim>(3));
   }
 
   {
+    constexpr int dim = 2;
     auto   mesh = import_mesh("beam-quad.mesh");
     Domain d0   = Domain::ofFaces(mesh, std::function([](std::vector<vec2> vertices, int /* attr */) {
                                   return average(vertices)[0] < 2.25;  // x coordinate of face center
@@ -236,12 +257,16 @@ TEST(domain, of_faces)
     Domain d3 = d0 & d1;
     EXPECT_EQ(d3.quad_ids_.size(), 2);
     EXPECT_EQ(d3.dim_, 2);
+
+    // check that by_attr compiles
+    Domain d4 = Domain::ofFaces(mesh, by_attr<dim>(3));
   }
 }
 
 TEST(domain, of_elements)
 {
   {
+    constexpr int dim = 3;
     auto   mesh = import_mesh("patch3D_tets_and_hexes.mesh");
     Domain d0   = Domain::ofElements(mesh, std::function([](std::vector<vec3> vertices, int /*bdr_attr*/) {
                                      return average(vertices)[0] < 0.7;  // x coordinate of face center
@@ -267,9 +292,13 @@ TEST(domain, of_elements)
     EXPECT_EQ(d3.tet_ids_.size(), 0);
     EXPECT_EQ(d3.hex_ids_.size(), 0);
     EXPECT_EQ(d3.dim_, 3);
+
+    // check that by_attr works
+    Domain d4 = Domain::ofElements(mesh, by_attr<dim>(3));
   }
 
   {
+    constexpr int dim = 2;
     auto   mesh = import_mesh("patch2D_tris_and_quads.mesh");
     Domain d0   = Domain::ofElements(
         mesh, std::function([](std::vector<vec2> vertices, int /* attr */) { return average(vertices)[0] < 0.45; }));
@@ -292,6 +321,9 @@ TEST(domain, of_elements)
     EXPECT_EQ(d3.tri_ids_.size(), 0);
     EXPECT_EQ(d3.quad_ids_.size(), 0);
     EXPECT_EQ(d3.dim_, 2);
+
+    // check that by_attr compiles
+    Domain d4 = Domain::ofElements(mesh, by_attr<dim>(3));
   }
 }
 

